### PR TITLE
HWKAPM-514 Enable publishers to be selected via the ServiceResolver, …

### PR DESCRIPTION
**Pull request by objectiser**
…to enable selection between alternative implementations based on environment variables, as in the client

### CODE REVIEW

**Comment by objectiser** [comment]
Reviewer notes:
* This change is to enable alternative event publisher implementation to be chosen (currently only JMS available, but soon also Kafka)
* The ServiceResolver mechanism, using the ServiceStatus API to determine availability, is used on the client side - so making these changes makes the server side consistent with the client in this respect
* The MDBs have been updated - although they still use CDI to resolve (as used in a JEE server context only), they have been changed to directly only search for JMS implementations of the publishers - so once an event starts being processed by a particular technology, it (and any events it derives) should only be processed within the scope of that technology

File: server/jms/src/main/java/org/hawkular/apm/server/jms/CommunicationDetailsDeriverMDB.java [hunk 1]
```diff
@@ -49,7 +48,7 @@
     private TracePublisherJMS tracePublisher;
 
     @Inject
-    private CommunicationDetailsPublisher communicationDetailsPublisher;
+    private CommunicationDetailsPublisherJMS communicationDetailsPublisher;
```
Comment (pilhuhn) [inline server/jms/src/main/java/org/hawkular/apm/server/jms/CommunicationDetailsDeriverMDB.java]: I am probably just not understanding this. If you inject this, you should have a producer somewhere that can give you the concrete type of `CommunicationDetailsPublisher`. Why do you then specify the concrete type here? Remember that JMS and MDB are api/abstractions over the concrete message oriented middleware or the data source creating the message. There are resource adapters other than JMS that can drive MDBs

Comment (objectiser) [inline server/jms/src/main/java/org/hawkular/apm/server/jms/CommunicationDetailsDeriverMDB.java]: The reason is mentioned in the reviewer notes - it is like there will be atleast one other concrete implementation of the CommunicationDetailsPublisher soon (Kafka) so wanted to ensure that within the MDB/JMS context, only the JMS publisher implementation is chosen.

**Comment by objectiser** [comment]
@pilhuhn Thanks Heiko

**Pull request merged**